lemur/certificates/service.py

```diff
@@ -1005,3 +1009,105 @@ def remove_destination_association(certificate, destination):
+    """
+    Finds all certificates expiring soon but are still being used for TLS at any domain with which they are associated.
+    Identified ports will then be persisted on the certificate_associations row for the given cert/domain combo.
+
+    Note that this makes actual TLS network calls in order to establish the "deployed" part of this check.
+    """
+    all_certs = defaultdict(dict)
+    for c in get_certs_for_expiring_deployed_cert_check(exclude):
+        domains_for_cert = find_and_persist_domains_where_cert_is_deployed(c, timeout_seconds_per_network_call)
+        if len(domains_for_cert) > 0:
+            all_certs[c] = domains_for_cert
+    database.commit()
+
+
+def get_certs_for_expiring_deployed_cert_check(exclude):
+    now = arrow.utcnow()
+    threshold_days = current_app.config.get("LEMUR_EXPIRING_DEPLOYED_CERT_THRESHOLD_DAYS", 14)
+    max_not_after = now + timedelta(days=threshold_days)
+
+    q = (
+        database.db.session.query(Certificate)
+        .filter(Certificate.not_after <= max_not_after)
+        .filter(Certificate.expired == false())
+        .filter(Certificate.revoked == false())
+        .filter(Certificate.in_rotation_window == true())
+    )
+
+    exclude_conditions = []
+    if exclude:
+        for e in exclude:
+            exclude_conditions.append(~Certificate.name.ilike("%{}%".format(e)))
+
+        q = q.filter(and_(*exclude_conditions))
+
+    return windowed_query(q, Certificate.id, 10000)
+
+
+def find_and_persist_domains_where_cert_is_deployed(certificate, timeout_seconds_per_network_call):
+    """
+    Checks if the specified cert is still deployed. Returns a list of domains to which it's deployed.
+
+    We use the serial number to identify that a certificate is identical. If there were multiple certificates
+    issued for the same domain with identical serial numbers, this could return a false positive.
+
+    Note that this checks *all* configured ports (specified in config LEMUR_PORTS_FOR_DEPLOYED_CERTIFICATE_CHECK)
+    for all the domains in the cert. If the domain is valid but the port is not, we have to wait for the connection
+    to time out, which means this can be quite slow.
+
+    :return: A dictionary of the form {'domain1': [ports], 'domain2': [ports]}
+    """
+    matched_domains = defaultdict(list)
+    # filter out wildcards, we can't check them
+    for cert_association in [assoc for assoc in certificate.certificate_associations if '*' not in assoc.domain.name]:
+        domain = cert_association.domain
+        matched_ports_for_domain = []
+        for port in current_app.config.get("LEMUR_PORTS_FOR_DEPLOYED_CERTIFICATE_CHECK", [443]):
+            try:
+                parsed_serial = parse_serial(get_certificate_via_tls(domain.name, port,
+                                                                     timeout_seconds_per_network_call))
+                if parsed_serial == int(certificate.serial):
```

Comment:
I wonder if we can do something valuable with the information about the retrieved certificate. 
For instance, search for the existing cert in Lemur and update it.

In case the cert is not in Lemur, likely the case for self-signed certs and instance certs, I am wondering if emitting the certificate in the logs via some interesting properties, like CN, issuer, expiration date, would be a ground to alert on expired certs. 

This might complicate the current PR, and we can think about it in the future.

Reply:
Yeah, that could make sense. I'll table it for now and we can consider additional enhancements in the future.